### PR TITLE
fix(wait_for_pods_readiness): make it actually respect nodes count

### PR DESCRIPTION
**Pull request by dkropachev**
https://trello.com/c/X298xmfP/2556-k8s-waitforpodsreadiness-does-not-respect-node-count

## PR pre-checks (self review)

- [x] I followed [KISS principle](https://en.wikipedia.org/wiki/KISS_principle) and [best practices](https://docs.google.com/document/d/1jihgOKb5iGRlD8_HQ92O0JbLk1kASUoZT23i_MXFSKI)
- [x] I didn't leave commented-out/debugging code
- [x] I added the relevant `backport` labels
- [ ] ~~New configuration option are added and documented (in `sdcm/sct_config.py`)~~
- [ ] ~~I have added tests to cover my changes (Infrastructure only - under `unit-test/` folder)~~
- [x] All new and existing unit tests passed (CI)
- [ ] ~~I have updated the Readme/doc folder accordingly (if needed)~~

### CODE REVIEW

File: sdcm/cluster_k8s/__init__.py [hunk 1]
```diff
@@ -438,11 +438,15 @@ def wait_for_init(self):
     def wait_for_pods_readiness(self, count: Optional[int] = None):
         if count is None:
             count = len(self.nodes)
-        for _ in range(count):
-            time.sleep(self.pod_readiness_delay)
-            self.k8s_cluster.kubectl(f"wait --timeout={self.pod_readiness_timeout}m --all --for=condition=Ready pod",
-                                     namespace=self.namespace,
-                                     timeout=self.pod_readiness_timeout*60+10)
+        if count is None:
+            count = len(self.nodes)
```
Comment (enaydanov) [inline sdcm/cluster_k8s/__init__.py:442]: Dup of previous 2 lines

**Review by enaydanov — CHANGES_REQUESTED** [review]
Don't understand the problem you tried to solve here, but you're introducing another one: the pods are starting consequently and there are points in this process when all started pods are ready, but next is not started yet.

It'll be great to add details about the problem.

My suggestion: move a check for "condition met" count inside the loop, and break if the count is equal to the requested count, add "else" clause to the loop with raising RuntimeError.

File: sdcm/cluster_k8s/__init__.py (hunk 1 above)
Comment (dkropachev) [reply sdcm/cluster_k8s/__init__.py]: fixed

**Review by dkropachev — COMMENTED** [review]
> Don't understand the problem you tried to solve here, but you're introducing another one: the pods are starting consequently and there are points in this process when all started pods are ready, but next is not started yet.

Problem:
When you spin cluster more than 3 nodes you have good chance to slip thru the crack when one node just become ready and next one is not started yet.

> My suggestion: move a check for "condition met" count inside the loop, and break if the count is equal to the requested count, add "else" clause to the loop with raising RuntimeError.

Looping is already done via timeout decorator, don't think we need yet another.

File: sdcm/cluster_k8s/__init__.py [hunk 2]
```diff
@@ -438,11 +438,13 @@ def wait_for_init(self):
...
-            time.sleep(self.pod_readiness_delay)
-            self.k8s_cluster.kubectl(f"wait --timeout={self.pod_readiness_timeout}m --all --for=condition=Ready pod",
-                                     namespace=self.namespace,
-                                     timeout=self.pod_readiness_timeout*60+10)
+        time.sleep(self.pod_readiness_delay)
+        result = self.k8s_cluster.kubectl(
+            f"wait --timeout={self.pod_readiness_timeout}m --all --for=condition=Ready pod",
+            namespace=self.namespace,
+            timeout=self.pod_readiness_timeout * 60 + 10)
+        if result.stdout.count('condition met') != count:
```
Comment (enaydanov) [inline sdcm/cluster_k8s/__init__.py:446]: There is another problem: you need to handle situation when we grow cluster. In that case count will be equal to number of added nodes, and count of "condition met" to total nodes.

**Review by enaydanov — CHANGES_REQUESTED** [review]

File: sdcm/cluster_k8s/__init__.py (hunk 2 above)
Comment (dkropachev) [reply sdcm/cluster_k8s/__init__.py:446]: Nice finding, fixed.

File: sdcm/cluster_k8s/__init__.py [hunk 3]
```diff
@@ -438,11 +438,13 @@ def wait_for_init(self):
...
+            raise RuntimeError(f'Not all nodes reported')
```
Comment (enaydanov) [inline sdcm/cluster_k8s/__init__.py:447]: Don't need `f` here.

**Review by enaydanov — APPROVED** [review]
LGTM, except small cosmetic issue.

File: sdcm/cluster_k8s/__init__.py (hunk 3 above)
Comment (dkropachev) [reply sdcm/cluster_k8s/__init__.py]: fixed

**Pull request merged**